### ISSUE TITLE
NNFusion v0.2 (Nov~Dec) Release Plan

### ISSUE THREAD

**Issue by AlisaChen98** [microsoft/nnfusion#105]
This is the draft plan for the upcoming iteration in Nov ~ Dec, it's a 5 week's iteration.

## Release Plan for V0.2
* Release Manager：@wenxcs
* Feature Freeze Date：2020/12/08
* Code Freeze Date
* Release Date
* Release Note

### Mechanism
* support training | 5 days | dev-owner:@mzmssg #114 ( review&test @jlxue ) ETA:2020/11/26
   - [x] Inference from PyTorch mnist model
   - [x] Training from PyTorch mnist model
   - [x] Inference from HuggingFace BERT
* support low-precision & mixed-precision, 5 days 
   * Support basic inference models with fp16 datatype. (VGG16) 
   - [x] Implement DataBuffer and refactor code relating to it | 5 days @Niupple #115 ( review&test @wenxcs ) ETA:2020/11/17
   - [x] Check on each compiling pass, to make sure everything goes fine with fp16 | 5 days @Cjkkkk  #116( review&test @wenxcs ) ETA:2020/11/17
* auto kernel tuner integration | 5 days |  @jlxue 
   * IR support | 3 days @xiayuqing0622 @jlxue @ghostplant #117 ETA:2020/11/13
     - [x] Dot 
     - [x] BatchNormInference 
     - [x] Sum 
     - [x] Pad 
     - [x] AvgPool
     - [x] Softmax
     - [x] BatchMatMul
     - [x] Convert
     - [x] GatherV2
     - [x] OneHot
     - [x] Elementwise: Divide,Multiply, Sigmoid, Tanh,Power,Subtract,Rsqrt,Divide,Relu6
   * Auto tuning 
   - [x] Install or deploy local Antares server(Intermediate Layer for python/binary interaction) | 3 days @wenxcs #118( review&test @jlxue ) ETA:2020/11/27
   - [x] Antares support for enabling kernel tuning, dump progress in comments | 1 day @ghostplant #119( review&test @jlxue ) ETA:2020/11/16
   - [x] Add a graph pass for support kernel tuning only, report tuning status | 1 day @jlxue #120( review&test @xiayuqing0622 ) ETA:2020/11/16
   * Kernel DB 
   - [x] Parse Antares results and inject into kernel DB | 5 days @xysmlx #121 ETA:2020/11/24
* parallel training support (via SuperScaler) | 8 days | @lynex  #123( review&test @xiayuqing0622 ) 
   - [x] v0.1 intergation | 5 days @lynex 
   - [x] document | 3 days @lynex 

### Refactor
* detect unsupported model & ops | 2 days |  @mzmssg #124 ETA:2020/11/12
   - [x] Check unsupport op type| 1 day @mzmssg 
   - [x] Check unsupport model (control flow)| 1 day @mzmssg 
* support block-fusion as default | 5 days | @xysmlx #125 ETA:2020/11/17
   - [x] fix memory allocation problem #61 | 1 day @xysmlx ( review&test @wenxcs ) ETA:2020/11/11
   - [x] automatic failure check #110 | 2 days @xysmlx ETA:2020/11/14
* code refactor | 5 days |  @wenxcs #128 ETA:2020/11/17
   - [x] Move thirdparty/ngraph back to src/nnfusion | 5 days @xiayuqing0622 @mzmssg @wenxcs 

### Frontend&Backend Support
* support CPU |  @guoshzhao 
   - [x] Update azure mirror download urls for thirdparty package

---
## Stretch Goals
* parallel training support (via SuperScaler) |  @lynex 
   - [ ] v0.2 new datatype support
* support block-fusion as default |  @xysmlx 
   - [x] Define and implement the interfaces between BlockFusion and kernel tuner
   - [x] End-to-end test with kernel tuner enabled
   - [x] Refactor kernel DB with the auto kernel tuner integration team
   - [ ] automatic active block check #50
* model support | pending
* offline inference(PAI)
   - [ ] Batch backet inference | 3 days @wenxcs 
   - [ ] Offline inference wrapper | 5 days @wenxcs 
* sub-graph substitution |  5days @wenxcs #127( review&test @jlxue @Niupple @xysmlx ) 
   - [ ] Graph match feature, by FSM: | @wenxcs 
   - [ ] Replacing current Pattern Match; | @wenxcs

**Comment by nnfbot** [comment]
Thanks for the report @AlisaChen98! I will look into it ASAP! (I'm a bot).

**Comment by AlisaChen98** [comment]
### Major Features

#### Mechanism update

* Support the use of python interface to accelerate the training and inference of PyTorch model
* Support low-precision and mixed-precision models, including fp16 and fp32
* Provide auto kernel tuner integration:
  * Support more than 10 ops in IR, including Dot, Sum, Pad, Convert and etc
  * Support auto tuning via [Antares](https://github.com/microsoft/antares)
* Support parallel training via [SuperScaler](https://github.com/microsoft/SuperScaler)

#### Others

* Update related documentations
* Small enhancement on user experience and bug fix

**Comment by AlisaChen98** [comment]
### 特性更新

#### 机制

* 支持使用Python接口为PyTorch模型的训练和推理加速
* 支持不同精度的模型的编译，包括fp16和fp32
* 提供自动的内核函数
* Support the use of python interface to accelerate the training and inference of PyTorch model
* Support low-precision and mixed-precision model compilation, including fp16 and fp32
* Provide auto kernel tuner integration:
  * Support Antares operator IR for 60+ ops, including Dot, Sum, Pad, Convert, etc.
  * Support auto tuning via [Antares](https://github.com/microsoft/antares)
* Support parallel training via [SuperScaler](https://github.com/microsoft/SuperScaler)
* Support kernel database

#### Others

* Update related documentations
* Small enhancement on user experience and bug fix
